File: tests/cases/check/layer-dependencies/csharp/inputs/src/Infrastructure/Sql.cs
using CodeConstraints.Rules;

using Domain;

namespace Infrastructure;

[Layer("infrastructure")]
public class SqlConnection
{
    public void Execute(string sql) { }
}

// infrastructure -> domain: explicitly allowed. Dependency inversion.
[Layer("infrastructure")]
public class SqlOrderRepository : IOrderRepository
{
    public void Save(Order order) { }
}

File: tests/cases/check/layer-dependencies/python/inputs/src/infrastructure/sql.py
from cdec_rules import layer

from domain.repository import OrderRepository


@layer("infrastructure")
class SqlConnection:
    def execute(self, sql: str) -> None:
        pass


@layer("infrastructure")
class SqlOrderRepository(OrderRepository):
    """infrastructure -> domain: explicitly allowed. Dependency inversion
    done right."""

    def save(self, order: "Order") -> None:
        pass
